STM32 ADC: support prescaler from RCC

gautierg-st opened the pull request:
The goal of this PR is to add the support of the ADC prescaler located in RCC for some series (F1, F3, N6, U3).
Though already supported in F1 and F3, it was through a specific property of the RCC. To avoid increasing the complexity on this end, the prescaler is configured the same way as for a kernel clock, in the ADC `clocks` property in device tree.
This allows to remove the RCC specificities of F1 and F3.

This led to changes in the STM32 clock defines due to the fact that the N6 prescaler register field is 8-bit long, so the macros needed to be reworked to make it fit.

Lastly, since we can now have up to three different clocks in the `clocks` properties, instead of relying on ordering, the `clock-names` property is mandatory for all ADC instances. Names have been added in existing dtsi, dts and overlay files, but it shall be redefined if a clock source or a prescaler is added.

gautierg-st committed:
include: dt-bindings: clock: stm32: rework STM32_DT_CLOCK_SELECT details

Reworks the way the STM32_DT_CLOCK_SELECT builds its elements. Instead of
taking a mask, it takes the MSB of the field. From the MSB and LSB, we
calculate the width of the field, and this width is then stored (instead
of the mask). This allows to gain space for higher values for the fields.

This larger space is necessary to add the selection of the ADC prescaler
on STM32N6 because it is an 8-bit long field.

The allowed width is from 1 to 8 (and internally stored as 0-7 to fit on
3 bits).

STM32_DT_CLKSEL_MASK_GET keeps the same name, since we still need the mask,
and returns the bitmask from the width with the BIT_MASK macro. Other
STM32_DT_CLKSEL_MASK_* macros are renamed with WIDTH.

All call to STM32_DT_CLOCK_SELECT are updated to reflect the change and use
a width instead of a mask.

This also fixes a few issues like STM32H7 MCO macros, and adds MCO_PRE for
STM32N6.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
include: dt-bindings: clock: stm32: add adc prescaler for f1, f3, n6 and u3

This commit adds the RCC configurations for ADC prescaler for STM32F1, F3,
N6 and U3.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: adc: stm32: make clock-names property required in adc binding

To easily differentiate between the different clocks that can be configured
in device tree, make their naming mandatory, and explicit what the expected
names are.

Add these names in all dtsi and dts files that need them.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: remove specific rcc compatible for stm32f1 and f3

Now that the ADC prescaler are set within the driver using the clock
system, the specific rcc compatibles for F1 and F3 are no longer useful.
Replace them with the standard one (from which they were derived).

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: clock: stm32: remove adc prescaler setting

Now that the ADC prescaler are set within the driver using the clock
system, remove the specific setting of the prescaler from the clock driver.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: bindings: adc: update binding description

Update STM32 ADC binding description now that the STM32F3 ADC asynchronous
prescaler is set through the clock property.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
doc: releases: update migration guide for stm32 adc

Update the 4.3 migration guide to include the change made on the STM32
ADC clocks.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32f3_clock.h:
```diff
@@ -66,4 +71,25 @@
+#define ADC_PRE_DIV_2		0
+#define ADC_PRE_DIV_4		1
+#define ADC_PRE_DIV_6		2
+#define ADC_PRE_DIV_8		3
```
Maybe a separate `stm32f37x_clock.h` (as done for other series) should be introduced. This would allow for more consistency across series.

The definitions in this file can be gated behind a symbol that only it defines such that it overrides them cleanly, i.e.:
```diff
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32F37X_CLOCK_H_
+ /* STM32F37x has specific ADC prescaler definitions */
...
```

gautierg-st replied:
To be consistent with the other files, I used `#undef` to undefine all previous ADC symbols.

mathieuchopstm reviewed (commented):
Maybe the `STM32_DT_CLOCK_SELECT` update should be done in a separate PR. Not too blocking for me though.

gautierg-st commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
Ok with changing the commit title, but disagree with your prototype proposition. The rework was needed to gain space, make the macro more compact so that N6 prescaler could fit into it. Width is only 3-bit long, using another bit_pos would be 5-bit long. Reg would have to be reduced (not currently blocking since no reg are that long, but still, it could be needed one day).

Besides, I don't think it's a hurdle to "calculate" the width of the bitfield from RefMan (most are under 4-bit long, that's pretty easy to read).

mathieuchopstm replied:
> Ok with changing the commit title, but disagree with your prototype proposition. The rework was needed to gain space, make the macro more compact so that N6 prescaler could fit into it. Width is only 3-bit long, using another bit_pos would be 5-bit long. Reg would have to be reduced (not currently blocking since no reg are that long, but still, it could be needed one day).

I am not suggesting to change the storage format but merely how the macro accepts parameters. When I said:
> `width` would be computed internally as `(high_bit - low_bit + 1)`

I meant that it would be computed internally *as part of the macro* (storage format is unchanged and contains `width`)

gautierg-st replied:
Oh, ok, my bad, I misunderstood your proposition.
I don't have strong opinions on either choice, but I'll wait for other opinions before making any potential changes. @etienne-lms, @erwango ?

erwango replied:
What would be the benefit of ` high_bit, low_bit` ? Readability ?

gautierg-st commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -15,6 +15,14 @@ properties:
+      Expected names are the following:
+      - "adcx" for the enabling the register clock (mandatory)
+      - "adc_ker" for configuring the kernel clock (if applicable)
+      - "adc_pre" for configuring the prescaler (if applicable)
```
It's also the case for the kernel clock (at least sometimes, and it can even be shared with DAC), so it should be mentioned too.

gautierg-st commented:
Rebased to fix conflicts, and comments taken into account.

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32f37x_clock.h:
```diff
@@ -0,0 +1,38 @@
+
+#include "stm32f3_clock.h"
+
+/* On STM32F37x the ADC prescaler is located in CFGR1 and the prescaler values are more limited */
```
Nit:
```suggestion
/* On STM32F37x, the ADC prescaler is located in CFGR1 and the prescaler values are more limited */
```

gautierg-st replied:
Done

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32f3_clock.h:
```diff
@@ -36,34 +36,53 @@
-#define RTC_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 8, BDCR_REG)
+#define RTC_SEL(val)		STM32_DT_CLOCK_SELECT((val), 9, 8, BDCR_REG)
+
+/* ADC prescaler division factor for all F3 but F37x */
```
Nit:
```suggestion
/* ADC prescaler division factor for all F3 except F37x */
```

gautierg-st replied:
Done

mathieuchopstm reviewed (commented):
I'll do a second pass to make sure no selector has been broken but otherwise LGTM, minus a few points

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32_common_clocks.h:
```diff
@@ -23,30 +23,31 @@
- * @note 'mask' range:   0x00~0x1F    [ 22 : 26 ]
- * @note 'val' range:    0x00~0x1F    [ 27 : 31 ]
+ * @note 'shift' range:  0~31         [ 16 : 20 ]
+ * @note 'width' range:  0~7          [ 21 : 23 ]
```
```suggestion
 * @note 'width' range:  0~7          [ 21 : 23 ]  Value encodes bit fields width minus 1
```
Would explicitly address extra comment from https://github.com/zephyrproject-rtos/zephyr/pull/97725/commits/e29ae439b835f26f1315533d1812f6ffb13670d6#r2451959811.

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32f10x_clock.h:
```diff
@@ -19,7 +19,7 @@
 /** CFGR1 devices */
 #undef MCO1_SEL /* Need to redefine generic F1 MCO_SEL for connectivity line devices. */
-#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 0xF, 24, CFGR1_REG)
+#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 27, 24, CFGR1_REG)
```
According o the refman, I think the right values here are:
```c
#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 26, 24, CFGR1_REG)
```
I think this should be fixed in a specific appended commit.

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32h5_clock.h:
```diff
@@ -68,67 +68,67 @@
+#define USART7_SEL(val)		STM32_DT_CLOCK_SELECT((val), 20, 18, CCIPR1_REG)
+#define USART8_SEL(val)		STM32_DT_CLOCK_SELECT((val), 23, 21, CCIPR1_REG)
+#define USART9_SEL(val)		STM32_DT_CLOCK_SELECT((val), 26, 24, CCIPR1_REG)
+#define USART10_SEL(val)	STM32_DT_CLOCK_SELECT((val), 30, 27, CCIPR1_REG)
```
```suggestion
#define USART10_SEL(val)	STM32_DT_CLOCK_SELECT((val), 29, 27, CCIPR1_REG)

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32h7rs_clock.h:
```diff
@@ -78,44 +78,43 @@
+#define I2C23_SEL(val)		STM32_DT_CLOCK_SELECT((val), 9, 8, D2CCIPR_REG)
+#define I2C1_I3C1_SEL(val)	STM32_DT_CLOCK_SELECT((val), 13, 12, D2CCIPR_REG)
+#define LPTIM1_SEL(val)		STM32_DT_CLOCK_SELECT((val), 18, 16, D2CCIPR_REG)
+#define FDCAN_SEL(val)		STM32_DT_CLOCK_SELECT((val), 24, 22, D2CCIPR_REG)
```
```suggestion
#define FDCAN_SEL(val)		STM32_DT_CLOCK_SELECT((val), 23, 22, D2CCIPR_REG)

gautierg-st replied:
Done

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -59,35 +59,35 @@
-#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 0xF, 24, CFGR_REG)
-#define MCO1_PRE(val)           STM32_DT_CLOCK_SELECT((val), 0x7, 28, CFGR_REG)
+#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 17, 24, CFGR_REG)
+#define MCO1_PRE(val)           STM32_DT_CLOCK_SELECT((val), 31, 28, CFGR_REG)
```
```suggestion
#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 27, 24, CFGR_REG)
#define MCO1_PRE(val)           STM32_DT_CLOCK_SELECT((val), 30, 28, CFGR_REG)

gautierg-st replied:
Done

gautierg-st commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -15,6 +16,16 @@ properties:
+      - "adc_ker" for configuring the kernel clock (if applicable)
+      - "adc_pre" for configuring the prescaler (if applicable)
+      Kernel clock and prescaler may be shared between several ADC instances.
+      In that case, it must be ensured that there aren't any conflicts.
```
Done

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -1550,14 +1553,23 @@ static int adc_stm32_set_clock(const struct device *dev)
 
 	if (clock_control_on(clk,
-		(clock_control_subsys_t) &config->pclken[0]) != 0) {
+		(clock_control_subsys_t) &config->pclken) != 0) {
```
Worth to move into a single line?

gautierg-st replied:
Done

etienne-lms reviewed (commented):
For commit "dts: bindings: adc: update binding description", could you mention stm32 in the header line, e.g.:
`dts: bindings: adc: update STM32 binding description`

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32f1_clock.h:
```diff
@@ -39,13 +39,13 @@
 
 /** CFGR1 devices */
-#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 0x7, 24, CFGR1_REG)
+#define MCO1_SEL(val)           STM32_DT_CLOCK_SELECT((val), 27, 24, CFGR1_REG)
```
Expected MSBit value here is 26 or 27?
From reading the reference manuals, it's 26.

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32f7_clock.h:
```diff
@@ -71,31 +71,31 @@
 /** @brief Dedicated clocks configuration register selection helpers */
 /** DKCFGR2 devices */
-#define USART1_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 0, DCKCFGR2_REG)
+#define USART1_SEL(val)		STM32_DT_CLOCK_SELECT((val), 4, 0, DCKCFGR2_REG)
```
Done

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32l5_clock.h:
```diff
@@ -59,34 +59,34 @@
+#define LPTIM1_SEL(val)		STM32_DT_CLOCK_SELECT((val), 19, 18, CCIPR_REG)
+#define LPTIM2_SEL(val)		STM32_DT_CLOCK_SELECT((val), 21, 20, CCIPR_REG)
+#define LPTIM3_SEL(val)		STM32_DT_CLOCK_SELECT((val), 23, 22, CCIPR_REG)
+#define FDCAN_SEL(val)		STM32_DT_CLOCK_SELECT((val), 22, 24, CCIPR_REG)
```
Done

gautierg-st commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -45,7 +45,7 @@ properties:
       - "ASYNC" : independent and asynchronous with the bus clock
       One of the two values may not apply to some series. Refer to the RefMan.
-      If an asynchronous clock is selected, a domain clock in the clock property
+      If an asynchronous clock is selected, a domain clock in the "clocks" property
```
Done

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32mp13_clock.h:
```diff
@@ -115,41 +115,41 @@
+#define I2C12_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 0, I2C12CKSELR_REG)
+#define I2C3_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 0, I2C345CKSELR_REG)
+#define I2C4_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 3, I2C345CKSELR_REG)
+#define I2C5_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 6, I2C345CKSELR_REG)
```
I apparently missed to update that one. Fixed.

gautierg-st committed:
drivers: adc: stm32: add support for rcc prescaler clock

Some series like F1, F3, N6 and U3 use an ADC prescaler defined in the RCC.
Instead of adding specific properties in the RCC driver, use the secondary
clock system to configure the prescaler.

The ADC driver now configures the clocks depending on their presence and
their name. Three clocks can be defined:
- the register clock (mandatory for all series)
- the kernel clock (depends on series)
- the prescaler value (depends on series)

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -49,7 +49,7 @@ properties:
       - "ASYNC" : independent and asynchronous with the bus clock
       One of the two values may not apply to some series. Refer to the RefMan.
-      If an asynchronous clock is selected, a domain clock in the clock property
+      If an asynchronous clock is selected, a kernel clock in the "clocks" property
```
I won't block but "domain clock" are all over the place and we need to be consistent. We'll need to fix this one way or the other.
(Note that we've already moved from "kernel" to "domain" once)

mathieuchopstm replied:
IMO we should stick to `kernel clock` as it is the name used in RefMan.

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -1534,33 +1536,27 @@ static int adc_stm32h7_setup_boost(const struct adc_stm32_cfg *config, ADC_TypeD
-			return -EIO;
-		}
+	/* Enable ADC clock source if applicable */
+	if ((config->has_pclken_ker) &&
```
Can remove inner parentheses.
Ditto below.

gautierg-st replied:
Done

etienne-lms commented:
LGTM but you need to rebase your series.

gautierg-st committed:
dts: arm: st: fix linter

Fix formatting errors reported by the dts linter.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented:
> @kylebonnici Seems linter is wrong here: https://github.com/zephyrproject-rtos/zephyr/actions/runs/18772147196/job/53558655897?pr=97725#step:13:176:. It reports failure and proposes the following:
> 
> ```diff
>  			dmas = <&gpdma1 0 7 (STM32_DMA_PERIPH_TX | STM32_DMA_16BITS |
> -					     STM32_DMA_PRIORITY_HIGH)>,
> +				STM32_DMA_PRIORITY_HIGH)>,
>  			       <&gpdma1 1 6 (STM32_DMA_PERIPH_RX | STM32_DMA_16BI
> ```

I've fixed the other issues reported by linter, but I'm waiting for feedback on this one.

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32_common_clocks.h:
```diff
@@ -23,30 +23,31 @@
- * @note 'val' range:    0x00~0x1F    [ 27 : 31 ]
+ * @note 'shift' range:  0~31         [ 16 : 20 ]
+ * @note 'width' range:  0~7          [ 21 : 23 ] Value encodes bit fields width minus 1
+ * @note 'val' range:    0x00~0x1F    [ 24 : 31 ]
```
I think that now:
```suggestion
 * @note 'val' range:    0x00~0xFF    [ 24 : 31 ]
```

mathieuchopstm reviewed (commented):
Otherwise LGTM
